### PR TITLE
Add batching support to `SymbolicOp` classes

### PR DESCRIPTION
**Pull request by AlbertMitjans**
- Add `ScalarSymbolicOp` class, which has:
  - A method to detect if batching is used.
  - A `matrix` method that takes into account broadcasting.
  - A `_matrix` abstract method to compute the matrix without broadcasting.

### CODE REVIEW

**Comment by AlbertMitjans** [comment]
[sc-31247]

File: pennylane/ops/op_math/adjoint.py
Comment (AlbertMitjans) [inline pennylane/ops/op_math/adjoint.py:249]: We don't need this method.

**Review by timmysilv — COMMENTED** [review]
looks awesome! no major concerns from me. Could you elaborate a bit more on how you concluded that you should make this abstract class that's a child of `SymbolicOp`? I'm really fine with this deep inheritance tree, but I'd just like to make sure we're doing the best thing here.

I hinted to this in the code comment, but should `_matrix` be a static method? Based on your implementation, it seems halfway towards it. I don't think it'd add any value tbh, so I'm just curious what you're thinking

should we test other interfaces? in general i'm not concerned, but there is a method explicitly checking for `torch` compatibility issues...

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:231]: Wait, now you're making me doubt... I believe my code is wrong:

```pycon
>>> scalar = [1]
>>> [scalar] if qml.math.size(scalar) == 1 else scalar
[[1]]
```

I should use a `qml.math.ndim` check here.

Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py]: Changed to `qml.math.ndim` 👍

**Comment by AlbertMitjans** [comment]
>  Could you elaborate a bit more on how you concluded that you should make this abstract class that's a child of SymbolicOp?

Right now there are 3 `SymbolicOp` that also have a scalar: `Exp`, `SProd` and `Pow`. When allowing batching, we have three possibilities:
1. base is batched
2. scalar is batched
3. both are batched (with the same batching dimension)

Thus when computing the matrix, we must take into account these 3 possibilities (add 3 `if` statements). Without the `ScalarSymbolicOp` class, we would need to check these 3 possibilities in every matrix method of every `SymbolicOp` that contains a scalar (3 for now, maybe more in the future?)

Moreover, all `ScalarSymbolicOp` children don't need to worry about batching; everything is taken care of by its parent class.

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py]: Changed to `@staticmethod` 👍

File: pennylane/ops/op_math/symbolicop.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/symbolicop.py:226]: Yep, it is equivalent to `self.base.batch_size is None`, and also to `self.batch_size is None` (no batching is used).

We should not raise an error in this case, we just need to return a non-broadcasted matrix.

File: tests/ops/op_math/test_pow_op.py
Comment (AlbertMitjans) [reply tests/ops/op_math/test_pow_op.py:356]: Nope! Added assert back 👍

**Comment by AlbertMitjans** [comment]
> should we test other interfaces? in general i'm not concerned, but there is a method explicitly checking for torch compatibility issues...

Good point! Will add checks with other interfaces

File: pennylane/ops/op_math/exp.py
Comment (timmysilv) [reply pennylane/ops/op_math/exp.py]: I am utterly mystified because I am about 99.9% sure I tested my hunch correctly before making this comment, but I'm definitely seeing [[1]] now... glad I said something!

File: pennylane/ops/op_math/adjoint.py
Comment (dwierichs) [reply pennylane/ops/op_math/adjoint.py:249]: I think there was an argument that this keeps the batch_size attribute up to date. @albi3ro am I misremembering something here?

File: pennylane/ops/op_math/composite.py
Comment (dwierichs) [inline pennylane/ops/op_math/composite.py:77]: This is mostly a change of "style", right?

File: pennylane/ops/op_math/exp.py
Comment (dwierichs) [inline pennylane/ops/op_math/exp.py:60]: Very nice :100:

File: pennylane/ops/op_math/exp.py
Comment (dwierichs) [inline pennylane/ops/op_math/exp.py:251]: For a broadcasted `scalar`, this checks the interface for each entry of the argument, right? Isn't it easier to do
```suggestion
        coeff_interface = math.get_interface(scalar)
        scalar = [self.scalar] if qml.math.ndim(self.scalar) == 0 else self.scalar
```

File: pennylane/ops/op_math/exp.py
Comment (dwierichs) [inline pennylane/ops/op_math/exp.py:286]: We definitely should implement this functionality somewhere central! It is required by virtually _all_ `parametric_ops` `compute_matrix` methods.

File: pennylane/ops/op_math/exp.py
Comment (dwierichs) [inline pennylane/ops/op_math/exp.py:322]: Out of curiosity: Above you check for broadcasting with `ndim`, here with `size`. I suppose the check with `size` here would declare a scalar of shape `(1,)` to be non-broadcasted? That would break with current convention in PennyLane, as far as I can tell.

File: pennylane/ops/op_math/prod.py
Comment (dwierichs) [inline pennylane/ops/op_math/prod.py:64]: ...because this is more user-facing? :)

File: pennylane/ops/op_math/symbolicop.py
Comment (dwierichs) [inline pennylane/ops/op_math/symbolicop.py:67]: Just wondering about this change? I suppose it never really was an option to not pass `base`? Or do we loose any abstract default functionality by this change?

File: pennylane/ops/op_math/symbolicop.py
Comment (dwierichs) [inline pennylane/ops/op_math/symbolicop.py:173]: Like mentioned above, this might break the convention on whether we call something with shape `(1,)` that is expected to be a scalar broadcasted or not.

File: pennylane/ops/op_math/symbolicop.py
Comment (dwierichs) [inline pennylane/ops/op_math/symbolicop.py:239]: I like the case separation, really neatly done! :)

File: tests/ops/op_math/test_sprod.py
Comment (dwierichs) [inline tests/ops/op_math/test_sprod.py:915]: Good catch! :tada:

File: tests/tests_passing_pylint.txt
Comment (dwierichs) [inline tests/tests_passing_pylint.txt:7]: Are the other modified test files very far away from being ok? :sweat_smile:

**Review by dwierichs — COMMENTED** [review]
Very nice to see this support coming in! :) 
I think the `ScalarSymbolicOp` is a very handy intermediate abstraction.

My only concern is regarding the convention of whether we call a value with shape `(1,)` where we expect a scalar broadcasted or not. The current convention considers this to be broadcasting, whereas this PR changes this in two (out of three, I think) places.

File: pennylane/ops/op_math/adjoint.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint.py:249]: The way I see it, any `SymbolicOp` wraps a PL operator that is already initialised, thus `self.base._check_batching(params)` is already called.

Maybe we need this when updating parameters dynamically...?

File: pennylane/ops/op_math/composite.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/composite.py:77]: Yep!

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:251]: The problem is that `math.get_interface` always returns "numpy" when the input is a list (see function docstring).

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:322]: You're right, we should use `ndim` everywhere 👍

Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:322]: Changed!

File: pennylane/ops/op_math/symbolicop.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/symbolicop.py:67]: All properties and methods depend on `base`, thus a `SymbolicOp` with `base=None` would be unusable.

@albi3ro any reason why this default value was added?

File: pennylane/ops/op_math/symbolicop.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/symbolicop.py]: Changed!

File: pennylane/ops/op_math/composite.py
Comment (dwierichs) [reply pennylane/ops/op_math/composite.py:77]: Very nice :)

File: pennylane/ops/op_math/exp.py
Comment (dwierichs) [reply pennylane/ops/op_math/exp.py:251]: Ah, but we anyways do not want to support broadcasting via lists, so I believe this would not be a major issue, and be consistent with other operations.

File: pennylane/ops/op_math/prod.py
Comment (dwierichs) [reply pennylane/ops/op_math/prod.py:64]: Nice catch :)

File: pennylane/ops/op_math/adjoint.py
Comment (dwierichs) [reply pennylane/ops/op_math/adjoint.py:249]: Yes, exactly. So when updating the parameters of `Adjoint`, they might be broadcasted, but relying on `base.batch_size` later on will then report the batching status of the original parameters of the base, right?
The fact that you could remove it without noticing a problem means we're not testing for this, I suppose? :grimacing:

File: tests/tests_passing_pylint.txt
Comment (dwierichs) [reply tests/tests_passing_pylint.txt:11]: Nice!

File: pennylane/ops/op_math/adjoint.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint.py:249]: Yes, we're definitely not testing it. But I don't know if we want to support updating the parameters, given that we are pushing towards having unmutable classes...

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:251]: Yep you're right. I added your suggestion 👍

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:285]: Moved in `ScalarSymbolicOp.matrix` method!

Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py]: If we want to generalise it even further I'd recommend doing it in a separate PR

Comment (dwierichs) [reply pennylane/ops/op_math/exp.py]: I agree, it's just something I've seen so often, and usually there is no case separation done. I'll start a discussion on this.

File: pennylane/ops/op_math/adjoint.py
Comment (dwierichs) [reply pennylane/ops/op_math/adjoint.py:249]: Mhm. fair enough. Would it be a major pain point to leave this in, though? Just to not introduce a super nichy but possibly breaking change?

Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint.py:249]: Yep, maybe it's better to leave this in

Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint.py]: Added back!

**Review by dwierichs — APPROVED** [review]
Maybe good to wait for second approval and for @albi3ro to comment on the open conversation, but LGTM

File: pennylane/ops/op_math/symbolicop.py
Comment (albi3ro) [reply pennylane/ops/op_math/symbolicop.py:69]: Can't remember.  `base` is definitely a mandatory argument, so we should indeed get rid of `None`.

File: pennylane/ops/op_math/adjoint.py
Comment (timmysilv) [inline pennylane/ops/op_math/adjoint.py:60]: not gonna pretend I know how to beat Sphinx... but this [isn't rendering correctly](https://xanaduai-pennylane--3597.com.readthedocs.build/en/3597/code/api/pennylane.adjoint.html). I'm guessing you just need an empty line between the english text and the pycon

File: pennylane/ops/op_math/__init__.py
Comment (timmysilv) [inline pennylane/ops/op_math/__init__.py:72]: I can't comment there but can you add `~ScalarSymbolicOp` to the "Symbolic Classes" section above? Line 52

File: pennylane/ops/op_math/symbolicop.py
Comment (timmysilv) [inline pennylane/ops/op_math/symbolicop.py:161]: why not ScalarSymbolicOp?

File: pennylane/ops/op_math/symbolicop.py
Comment (timmysilv) [inline pennylane/ops/op_math/symbolicop.py:228]:
```suggestion
            # otherwise get `RuntimeError: Can't call numpy() on Tensor that requires grad.`
```

**Review by timmysilv — COMMENTED** [review]
did a lil docs sweep

File: pennylane/ops/op_math/adjoint.py
Comment (AlbertMitjans) [inline pennylane/ops/op_math/adjoint.py:59]:
```suggestion
        But it doesn't support batching of operators:

        >>> op = qml.adjoint([qml.RX(1, wires=0), qml.RX(2, wires=0)])
        ValueError: The object [RX(1, wires=[0]), RX(2, wires=[0])] of type <class 'list'> is not callable.
```

File: pennylane/ops/op_math/adjoint.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint.py:61]: Yep I believe you're right

File: pennylane/ops/op_math/symbolicop.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/symbolicop.py:161]: Oops, changed 👍

File: pennylane/ops/op_math/__init__.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/__init__.py:72]: Added!

File: pennylane/ops/op_math/adjoint.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint.py:61]: Now it is rendering correctly!

**Review by timmysilv — APPROVED** [review]

**Pull request merged**
